### PR TITLE
Implement SwapLimit for Ringswap

### PR DESCRIPTION
**Pull request by sunspirit99**

### CODE REVIEW

File: pkg/liquidity-source/ringswap/pool_simulator.go [hunk 1]
```diff
@@ -85,56 +91,30 @@ func (s *PoolSimulator) CalcAmountOut(param poolpkg.CalcAmountOutParams) (*poolp
...
-	}
-
-	originalReserve, overflow := uint256.FromBig(lo.Ternary(reserveOutIndex == 0, s.originalReserves.Reserve0, s.originalReserves.Reserve1))
-	if overflow {
-		return nil, uniswapv2.ErrInvalidReserve
-	}
-
-	// Ensure that amountOut does not exceed the original reserve
-	if amountOut.Cmp(originalReserve) > 0 {
+	if amountOut.Cmp(reserveOut) >= 0 {
```
Comment (NgoKimPhu) [inline pkg/liquidity-source/ringswap/pool_simulator.go:100]: Can it not be equal?

File: pkg/liquidity-source/ringswap/pool_simulator.go [hunk 2]
```diff
@@ -85,56 +91,30 @@ func (s *PoolSimulator) CalcAmountOut(param poolpkg.CalcAmountOutParams) (*poolp
...
 
-	wTokenOutIndex := indexOut%2 + 2
-	if wTokenOutIndex >= len(s.Pool.Info.Tokens) {
-		return nil, ErrTokenIndexOutOfBounds
+	// Ensure that amountOut does not exceed original reserve
+	if isUnwrapOut {
+		if param.Limit == nil {
+			return nil, ErrNoSwapLimit
+		}
+		if amountOut.CmpBig(param.Limit.GetLimit(wTokenOut)) >= 0 {
```
Comment (NgoKimPhu) [inline pkg/liquidity-source/ringswap/pool_simulator.go:114]: Same here

File: pkg/liquidity-source/ringswap/pool_simulator.go [hunk 3]
```diff
@@ -269,12 +231,35 @@ func (s *PoolSimulator) UpdateBalance(params poolpkg.UpdateBalanceParams) {
...
+	if !ok {
+		return
+	}
+
 	if indexIn%2 == 0 {
-		s.originalReserves.Reserve0 = new(big.Int).Add(s.originalReserves.Reserve0, params.TokenAmountIn.Amount)
-		s.originalReserves.Reserve1 = new(big.Int).Sub(s.originalReserves.Reserve1, params.TokenAmountOut.Amount)
+		if params.SwapLimit != nil {
+			if swapInfo.IsWrapIn {
+				s.originalReserves.Reserve0 = new(big.Int).Add(s.originalReserves.Reserve0, params.TokenAmountIn.Amount)
```
Comment (NgoKimPhu) [inline pkg/liquidity-source/ringswap/pool_simulator.go:242]: Now that we're using swap limit, can we remove this?

File: pkg/liquidity-source/ringswap/pool_simulator.go (hunk 1 above)
Comment (sunspirit99) [reply pkg/liquidity-source/ringswap/pool_simulator.go:100]: yes, it must be gt : 
`require(amount0Out < _reserve0 && amount1Out < _reserve1, 'SwapV2: INSUFFICIENT_LIQUIDITY');`

File: pkg/liquidity-source/ringswap/pool_simulator.go (hunk 2 above)
Comment (sunspirit99) [reply pkg/liquidity-source/ringswap/pool_simulator.go:114]: yes, it must be gt :
`require(amount0Out < _reserve0 && amount1Out < _reserve1, 'SwapV2: INSUFFICIENT_LIQUIDITY');`

File: pkg/liquidity-source/ringswap/pool_simulator.go [hunk 4]
```diff
@@ -269,12 +231,35 @@ func (s *PoolSimulator) UpdateBalance(params poolpkg.UpdateBalanceParams) {
...
+				_, _, _ = params.SwapLimit.UpdateLimit("", swapInfo.WTokenIn, bignumber.ZeroBI, params.TokenAmountOut.Amount)
+			}
+
+			if swapInfo.IsUnwrapOut {
+				s.originalReserves.Reserve1 = new(big.Int).Sub(s.originalReserves.Reserve1, params.TokenAmountOut.Amount)
+				_, _, _ = params.SwapLimit.UpdateLimit(swapInfo.WTokenOut, "", params.TokenAmountIn.Amount, bignumber.ZeroBI)
```
[the remark this answers is not in the document]
Comment (sunspirit99) [reply pkg/liquidity-source/ringswap/pool_simulator.go]: now, it looks more concise than ever  :) : https://github.com/KyberNetwork/kyberswap-dex-lib/pull/610/commits/00c92c66da2934663cd75ecce647ad2a98070680

File: pkg/liquidity-source/ringswap/pool_simulator.go (hunk 3 above)
Comment (sunspirit99) [reply pkg/liquidity-source/ringswap/pool_simulator.go]: i removed it : 
https://github.com/KyberNetwork/kyberswap-dex-lib/pull/610/commits/00c92c66da2934663cd75ecce647ad2a98070680

File: pkg/liquidity-source/ringswap/pool_simulator.go [hunk 5]
```diff
@@ -269,12 +230,20 @@ func (s *PoolSimulator) UpdateBalance(params poolpkg.UpdateBalanceParams) {
...
+
+	if params.SwapLimit != nil {
+		if swapInfo.IsWrapIn {
+			amount := lo.Ternary(indexIn%2 == 0, params.TokenAmountOut.Amount, params.TokenAmountIn.Amount)
+			_, _, _ = params.SwapLimit.UpdateLimit("", swapInfo.WTokenIn, bignumber.ZeroBI, amount)
+		}
+		if swapInfo.IsUnwrapOut {
+			amount := lo.Ternary(indexIn%2 == 0, params.TokenAmountIn.Amount, params.TokenAmountOut.Amount)
+			_, _, _ = params.SwapLimit.UpdateLimit(swapInfo.WTokenOut, "", amount, bignumber.ZeroBI)
+		}
```
Comment (NgoKimPhu) [inline pkg/liquidity-source/ringswap/pool_simulator.go:246]: do something like this:
```
amtIn, amtOut := params.TokenAmountIn.Amount,  params.TokenAmountOut.Amount
if indexIn%2 == 0 { // double check this, why do we have this check?
  amtIn, amtOut = amtOut, amtIn
}
updateIn, deltaIn := "", bignumber.ZeroBI
if swapInfo.IsWrapIn {
  updateIn, deltaIn = swapInfo.WTokenIn, amtIn
}
updateOut, deltaOut := "", bignumber.ZeroBI
if swapInfo.IsUnwrapOut {
  updateOut, deltaOut = swapInfo.WTokenOut, amtOut
}

_, _, _ = params.SwapLimit.UpdateLimit(updateOut, updateIn, deltaOut, deltaIn)

Comment (sunspirit99) [reply pkg/liquidity-source/ringswap/pool_simulator.go]: https://github.com/KyberNetwork/kyberswap-dex-lib/pull/610/commits/ea8afe3b5c5219330b1392479a5702793c120f5a

**Review by NgoKimPhu — APPROVED** [review]

**Pull request merged**